File: hnsx-server/pkg/spec/presets.go
package spec

// PresetName is a named policy bundle that expands into concrete budget,
// permission, approval and guardrail settings.
type PresetName string

const (
	// SafeCustomerService blocks destructive side-effects and requires human
	// approval for refunds or customer-data exports.
	SafeCustomerService PresetName = "safe_customer_service"
)

// PresetRegistry maps preset names to their expanded PolicySpec. Presets are
// merged in order; explicit user values override preset defaults.
var PresetRegistry = map[PresetName]PolicySpec{
	SafeCustomerService: {
		Budget: BudgetSpec{
			MaxCostUSD: 1.0,
			MaxTurns:   20,
		},
		Permissions: PermissionSpec{
			AllowNetwork:    true,
			AllowFileWrite:  false,
			AllowFileDelete: false,
			AllowShell:      false,
		},
		Approval: ApprovalSpec{
			DefaultTimeoutSeconds: 600,
			RequiredFor: RequiredForSpec{
				Tools:     []string{"issue_refund", "export_customer_data"},
				Resources: []string{"billing:write", "customer:*"},
			},
		},
		Guardrails: []GuardrailSpec{
			{
				ID:      "block-pii-keywords",
				Type:    "output_keyword",
				On:      "output",
				Action:  "block",
				Message: "output contains potential PII",
				Config:  map[string]any{"keywords": []string{"password", "credit card", "ssn"}},
			},
		},
	},
}

// ExpandPresets returns a new PolicySpec with all named presets merged into
// the base spec. User-provided fields take precedence over preset defaults.
func ExpandPresets(base PolicySpec) PolicySpec {
	out := base
	for _, name := range base.Presets {
		preset, ok := PresetRegistry[PresetName(name)]
		if !ok {
			continue
		}
		mergePreset(&out, preset)
	}
	return out
}

func mergePreset(dst *PolicySpec, src PolicySpec) {
	if dst.Budget.MaxCostUSD == 0 {
		dst.Budget.MaxCostUSD = src.Budget.MaxCostUSD
	}
	if dst.Budget.MaxTurns == 0 {
		dst.Budget.MaxTurns = src.Budget.MaxTurns
	}
	if dst.Budget.MaxTokens == 0 {
		dst.Budget.MaxTokens = src.Budget.MaxTokens
	}

	// Permission booleans: explicit false is meaningful, so only override when
	// the dst value is the zero value (false). In practice this means a preset
	// can turn a capability ON, but turning it OFF requires an explicit user
	// value; this matches the "safe" preset semantics.
	if !dst.Permissions.AllowFileWrite && src.Permissions.AllowFileWrite {
		dst.Permissions.AllowFileWrite = true
	}
	if !dst.Permissions.AllowFileDelete && src.Permissions.AllowFileDelete {
		dst.Permissions.AllowFileDelete = true
	}
	if !dst.Permissions.AllowNetwork && src.Permissions.AllowNetwork {
		dst.Permissions.AllowNetwork = true
	}
	if !dst.Permissions.AllowShell && src.Permissions.AllowShell {
		dst.Permissions.AllowShell = true
	}

	if dst.Approval.DefaultTimeoutSeconds == 0 {
		dst.Approval.DefaultTimeoutSeconds = src.Approval.DefaultTimeoutSeconds
	}
	dst.Approval.RequiredFor.Tools = mergeUnique(dst.Approval.RequiredFor.Tools, src.Approval.RequiredFor.Tools)
	dst.Approval.RequiredFor.Resources = mergeUnique(dst.Approval.RequiredFor.Resources, src.Approval.RequiredFor.Resources)
	if dst.Approval.RequiredFor.CostThresholdUSD == 0 {
		dst.Approval.RequiredFor.CostThresholdUSD = src.Approval.RequiredFor.CostThresholdUSD
	}

	dst.Guardrails = append(dst.Guardrails, src.Guardrails...)
}

func mergeUnique(a, b []string) []string {
	seen := make(map[string]struct{}, len(a)+len(b))
	out := make([]string, 0, len(a)+len(b))
	for _, v := range a {
		if _, ok := seen[v]; ok {
			continue
		}
		seen[v] = struct{}{}
		out = append(out, v)
	}
	for _, v := range b {
		if _, ok := seen[v]; ok {
			continue
		}
		seen[v] = struct{}{}
		out = append(out, v)
	}
	return out
}

// KnownPresets returns every registered preset name.
func KnownPresets() []string {
	out := make([]string, 0, len(PresetRegistry))
	for name := range PresetRegistry {
		out = append(out, string(name))
	}
	return out
}

File: hnsx-worker/hnsx_worker/policy/presets.py
"""Policy presets — named bundles that expand into concrete policy config.

Presets keep the DomainSpec YAML concise for common scenarios while the
runtime still sees fully expanded budget / permission / approval / guardrail
values.
"""

from __future__ import annotations

from typing import Any

PRESET_REGISTRY: dict[str, dict[str, Any]] = {
    "safe_customer_service": {
        "budget": {
            "max_cost_usd": 1.0,
            "max_turns": 20,
        },
        "permissions": {
            "allow_network": True,
            "allow_file_write": False,
            "allow_file_delete": False,
            "allow_shell": False,
        },
        "approval": {
            "default_timeout_seconds": 600,
            "required_for": {
                "tools": ["issue_refund", "export_customer_data"],
                "resources": ["billing:write", "customer:*"],
            },
        },
        "output_guardrails": {
            "blocked_keywords": ["password", "credit card", "ssn"],
        },
    },
}


def known_presets() -> list[str]:
    """Return every registered preset name."""
    return list(PRESET_REGISTRY.keys())


def expand_presets(policy: dict[str, Any]) -> dict[str, Any]:
    """Return a new policy dict with named presets merged in.

    User-provided values take precedence over preset defaults. Multiple presets
    are merged in order; later presets override earlier ones.

    Raises:
        ValueError: if a preset name is not registered.
    """
    out: dict[str, Any] = {k: v for k, v in policy.items() if k != "presets"}
    presets = policy.get("presets") or []
    for name in presets:
        preset = PRESET_REGISTRY.get(name)
        if preset is None:
            raise ValueError(
                f"unknown policy preset {name!r} (known: {', '.join(known_presets())})"
            )
        out = _merge_policy(out, preset)
    return out


def _merge_policy(dst: dict[str, Any], src: dict[str, Any]) -> dict[str, Any]:
    """Merge ``src`` preset defaults into ``dst`` (user values win)."""
    out: dict[str, Any] = {}
    for key in set(dst) | set(src):
        dst_val = dst.get(key)
        src_val = src.get(key)
        if isinstance(dst_val, dict) and isinstance(src_val, dict):
            out[key] = _merge_dict(dst_val, src_val)
        elif isinstance(dst_val, list) and isinstance(src_val, list):
            out[key] = _merge_list(dst_val, src_val)
        elif dst_val is not None and dst_val != {} and dst_val != []:
            out[key] = dst_val
        else:
            out[key] = src_val
    return out


def _merge_dict(dst: dict[str, Any], src: dict[str, Any]) -> dict[str, Any]:
    out = dict(src)
    for k, v in dst.items():
        if isinstance(v, dict) and isinstance(out.get(k), dict):
            out[k] = _merge_dict(v, out[k])
        elif isinstance(v, list) and isinstance(out.get(k), list):
            out[k] = _merge_list(v, out[k])
        elif v is not None and v != {} and v != []:
            out[k] = v
    return out


def _merge_list(dst: list[Any], src: list[Any]) -> list[Any]:
    seen = set()
    out: list[Any] = []
    for item in dst + src:
        key = _item_key(item)
        if key in seen:
            continue
        seen.add(key)
        out.append(item)
    return out


def _item_key(item: Any) -> Any:
    """Return a hashable key for deduplicating merged list items."""
    if isinstance(item, dict):
        return tuple(sorted((k, _item_key(v)) for k, v in item.items()))
    if isinstance(item, list):
        return tuple(_item_key(v) for v in item)
    return item
